### DISCUSSION TITLE
Models seem to stop working with Tools in the latest update 0.3.32

### DISCUSSION THREAD

**Discussion by makelvin2010** [open-webui/open-webui#6079] [General]
I was able to successfully use Tools with different Models; it seems to work particular well with Qwen 2.5. I was able to get it use all of the Tools that I made available to the Model. After upgrading to the latest version of WebUI 0.3.32, it just can seem to be able to access any of the Tools that I made available to it. The only thing that it seems to be able to access is Web Search which seems to be a builtin tool with WebUI now. Previously, I have to import the Web Search tool manually for it to do the web search through my local SearXNG.

Unfortunately, I didn't update my WebUI that frequently and I don't remember the last version that I have all of the Tools working properly before proceeding to update with the latest version 0.3.32. So I can't saying exactly at which version it stopped working. I believe the last time I updated was a couple of months ago and vaguely remember it might be something like 0.3.12 or 0.3.14. I am just not sure. 

I tried to download version 0.3.14 from github and unzip the source file into a directory; but when I go into the directory and typed "docker compose up -d" after I removed the 0.3.32 from the docker image, I still end up with 0.3.32 again instead of 0.3.14. I am not sure how I can get back to the older version. Can someone please help me downgrade to an older version of WebUI or hopefully fix the Tools issues with the latest 0.3.32? I really would like to use the Tools for my Models again.

**Comment by makelvin2010** [comment]
So far, I've tried all the basic tools like calculator, weather, and date and time. They all seems to be working properly in the Models. Unfortunately, the tools like the Web Search (one that is not built-in) and the Yahoo Finance are not working. Whenever I ask the Model to use these tools to retrieve the required information, it just simply say that it does not have access to these tools. Since the Web Search tool is now built into the WebUI itself, technically, I don't actually need to have that one working; but the Yahoo Finance tool is the one I truly need to have it going. 

Since both of these tools stopped working and get recognized by the Models at the same time, I wonder if there are some commonality between them that this new version of WebUI is preventing them from working. Is there a new hidden switch somewhere I need to enabled or disabled for this to work. I can't help thinking that the built-in Web Search capability might have something to do with it so I go and have that feature disabled. Unfortunately, nothing's changed. When I have the built-in Web Search enabled, whenever I try to ask for stock information, it will always use the Web Search to try to get the result instead of Yahoo Finance tool. If I have the built-in Web Search disabled, Yahoo Finance tool still does not work and the external Web Search tool also won't work as well.

**Comment by makelvin2010** [comment]
Looks like I have to modify the comments in the Tools functions to make it more explicit in using the tools function before I can finally get the Models to actually use the tools. Strange why that none of the Model files have changed; but its behavior on how it reacts to the comments have changed by changing and updating WebUI.

**Comment by CRCODE22** [comment]
Qwen 2.5 32B is not seeing and using any of the tools in OpenWebui even tough they are all enabled for it.

**Comment by CRCODE22** [comment]
I have also tried llama3.1 and this does not see tools either in Open Webui. It looks like for local models tools no longer work

**Comment by lee-b** [comment]
Agree, this isn't working for me either. Tried multiple local models and high-end service provider models. I've also noticed that the list of tools doesn't update if you update them on the toolserver side, even if you disable, re-enable, and re-test the service on the open-webui side. I'd say it's just broken right now -- something to do with caching perhaps.

**Comment by digitalassassins** [comment]
Ive written a tool to format telephone numbers and no matter how hard I try, Gemma, Llama etc, don't want to use the tool correctly. They have identified the tool, each time I call them out to use it. They simply ignore it and do their own thing. Ive put explicit system prompts and explicit comments to force usage but it's just not working. This might be why. if everybody else is having the same issue.

**Reply by Classic298** [reply]
what model do you use exactly? Small models are very very bad with function calling and tool use.

**Comment by digitalassassins** [comment]
It keeps saying it is in a frustrating loop and can't run the number through the Python script. It just regurgitates what I've already told it earlier on in the chat window. It gave me an example number 555-1234567 and asked me how the Python script would format this number using the tool. Now, whenever I pass it a telephone number to format. It just copies and pastes "Number: 5  5  5  1  2  3  4  5  6" 😂 
Gemma & Llama mainly. But will try with different models to see if I can get any to work.
![2025-05-15 19_50_47-🐍 Python String Literal • Open WebUI and 1 more page - Personal - Microsoft​ Ed](https://github.com/user-attachments/assets/6959d19e-994d-4302-a303-2667a6aadc08)

**Comment by filmo** [comment]
Are there any tips/trick to set up OpenWebUI to 'force/highly suggest' tool usage (for example, initial model prompts) that say essentially. "Examine tools available closely and give preference to them at all times unless specifically stated otherwise?"

For example on both DeepSeek R1 70b and llama3.3 70b, if I enter something like "What is 123/1234 * 456?" it will often ignore the Calculator Tool I have installed. If I explicitly say "Use the calculator tool" it will often (but not always) use the tool to get the precisely correct answer. If I follow up with additional math problems, it will hit-or-miss use the calculator tool. 

My calculator tool uses sympy which handles a wide range of evaluations:

```
import sympy as sp

class Tools:
    def __init__(self):
        pass

    def calculator(self, equation: str) -> str:
        """
        Calculate the result of a math equation safely and accurately.
        :param equation: The equation to calculate.
        :return: The result of the equation.
        """
        try:
            # Parse the equation using sympy
            expr = sp.sympify(equation)
            result = expr.evalf()
            return f"{equation} = {result}"
        except (sp.SympifyError, ValueError) as e:
            print(e)
            return "Invalid equation"
```

What I'm looking for is a way in setting to hint/suggest/force the models to rely on tools without having to preface every prompt with "Use tool X Y Z"
